File: mac/LRActionKit/LRActionKit/Action/LRContextAction.m
@import LRCommons;
@import PiiVersionKit;
@import PackageManagerKit;

#import "LRContextAction.h"
#import "LRActionManifest.h"
#import "LRVersionSpec.h"
#import "LRActionKit-Swift.h"


NSString *const LRContextActionDidChangeVersionsNotification = @"LRContextActionDidChangeVersions";


@interface LRContextAction ()

@end


@implementation LRContextAction {
    ATCoalescedState _updateState;
}

- (id)initWithAction:(Action *)action project:(id<ProjectContext>)project resolutionContext:(LRPackageResolutionContext *)resolutionContext {
    self = [super init];
    if (self) {
        _action = action;
        _project = project;
        _resolutionContext = resolutionContext;
        NSLog(@"%@ created", self);

        [self observeNotification:LRPackageContainerDidChangePackageListNotification withSelector:@selector(_updateAvailableVersions)];
        [self _updateAvailableVersions];
    }
    return self;
}

- (void)dealloc {
    [self removeAllObservations];
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ #ver=%d proj=%@", _action, (int)_versions.count, _project];
}

- (void)_updateAvailableVersions {
    AT_dispatch_coalesced_with_notifications(&_updateState, 0, ^(dispatch_block_t done) {
        _versions = [[self _computeAvailableVersions] copy];
        _versionSpecs = [[self _computeAvailableVersionSpecs] copy];
        [self postNotificationName:LRContextActionDidChangeVersionsNotification];
        done();
    }, ^{
        NSLog(@"ContextAction(%@) _updateState=%d", _action.name, (int)_updateState);
        [_project setAnalysisInProgress:(_updateState > 0) forTask:self];
    });
}

- (NSArray *)_computeAvailableVersions {
    if (!_action.valid) {
        NSLog(@"ContextAction(%@) has invalid manifest, so no versions", _action.name);
        return @[];
    }

    NSMutableArray *packageSets = [NSMutableArray new];
    for (LRAssetPackageConfiguration *configuration in _action.packageConfigurations) {
        [packageSets addObjectsFromArray:[_resolutionContext packageSetsMatchingPackageReferences:configuration.packageReferences]];
    }

    NSMutableArray *versions = [NSMutableArray new];
    for (LRPackageSet *packageSet in packageSets) {
        LRActionManifest *manifest = [self _actionManifestForPackageSet:packageSet];
        if (!manifest.valid) {
            NSLog(@"ContextAction(%@) skipping version %@ b/c of invalid manifest: %@", _action.name, packageSet.primaryPackage, manifest.errors);
        } else {
            LRActionVersion *version = [[LRActionVersion alloc] initWithAction:_action manifest:manifest packageSet:packageSet];
            [versions addObject:version];
        }
    }
    [versions sortUsingComparator:^NSComparisonResult(LRActionVersion *obj1, LRActionVersion *obj2) {
        return [obj1.primaryVersion compare:obj2.primaryVersion];
    }];
    NSLog(@"ContextAction(%@) has %d versions", _action.name, (int)versions.count);
    return versions;
}

- (NSArray *)_computeAvailableVersionSpecs {
    NSMutableArray *specs = [NSMutableArray new];
    NSMutableSet *set = [NSMutableSet new];

    void (^addSpec)(LRActionVersion *actionVersion, LRVersionSpec *versionSpec) = ^(LRActionVersion *actionVersion, LRVersionSpec *versionSpec) {
        if (![set containsObject:versionSpec]) {
            versionSpec.changeLogSummary = actionVersion.manifest.changeLogSummary;
            [set addObject:versionSpec];
            [specs addObject:versionSpec];
        }
    };

    for (LRActionVersion *actionVersion in _versions) {
        addSpec(actionVersion, [LRVersionSpec stableVersionSpecWithMajorFromVersion:actionVersion.primaryVersion]);
        addSpec(actionVersion, [LRVersionSpec versionSpecMatchingMajorMinorFromVersion:actionVersion.primaryVersion]);
        addSpec(actionVersion, [LRVersionSpec versionSpecMatchingVersion:actionVersion.primaryVersion]);
    }

    [specs addObject:[LRVersionSpec stableVersionSpecMatchingAnyVersionInVersionSpace:_action.primaryVersionSpace]];

    return specs;
}

- (void)_enumerateVersionSpecsForVersion:(LRActionVersion *)actionVersion block:(void(^)(LRVersionSpec *versionSpec))addSpec {
}

- (LRActionManifest *)_actionManifestForPackageSet:(LRPackageSet *)packageSet {
    NSMutableArray *layers = [NSMutableArray new];
    for (LRManifestLayer *layer in _action.manifestLayers) {
        if ([packageSet matchesAllPackageReferencesInArray:layer.packageReferences]) {
            [layers addObject:layer];
        }
    }
    return [[LRActionManifest alloc] initWithLayers:layers];
}

- (Rule *)newInstanceWithMemento:(NSDictionary *)memento {
    return [_action newRuleWithContextAction:self memento:memento];
}

@end

File: LRActionKit/Source/Action/LRContextAction.swift
import Foundation
import ExpressiveFoundation
import ExpressiveCollections
import PackageManagerKit
import Swift

public class LRContextAction: NSObject {

    public static let didChangeVersionsNotification = "LRContextActionDidChangeVersions"
    
    public let action: Action
    public let project: ProjectContext
    public let resolutionContext: LRPackageResolutionContext

    public private(set) var versions: [LRActionVersion] = []
    public private(set) var versionSpecs: [LRVersionSpec] = []

    public init(action: Action, project: ProjectContext, resolutionContext: LRPackageResolutionContext) {
        self.action = action
        self.project = project
        self.resolutionContext = resolutionContext
        super.init()

        updateCoalescence.monitorBlock = { (running) in 
            project.setAnalysisInProgress(running, forTask: self)
        }

        o.on(LRPackageContainerDidChangePackageListNotification, self, LRContextAction.updateAvailableVersions)
        updateAvailableVersions()
    }

    public override var description: String {
        return "\(action) #ver=\(versions.count) proj=\(project)"
    }

    public func newInstance(memento memento: [String: AnyObject]) -> Rule {
        return action.newRule(contextAction: self, memento: memento)
    }

    private func updateAvailableVersions() {
        updateCoalescence.perform {
            self.versions = self._computeAvailableVersions()
            self.versionSpecs = self._computeAvailableVersionSpecs()
            self.postNotification(LRContextAction.didChangeVersionsNotification)
        }
    }

    private func _computeAvailableVersions() -> [LRActionVersion] {
        if !action.valid {
            print("ContextAction(\(action.name)) has invalid manifest, so no versions")
            return []
        }

        let packageSets = action.packageConfigurations.map { self.resolutionContext.packageSetsMatchingPackageReferences($0.packageReferences) as! [LRPackageSet] }.flatten()

        let versions = packageSets.mapIf { packageSet -> LRActionVersion? in
            let layers = self.action.manifestLayers.filter { packageSet.matchesAllPackageReferencesInArray($0.packageReferences) }
            let manifest = LRActionManifest(layers: layers)
            if manifest.valid {
                return LRActionVersion(action: self.action, manifest: manifest, packageSet: packageSet)
            } else {
                print("ContextAction(\(self.action.name)) skipping version \(packageSet.primaryPackage) b/c of invalid manifest: \(manifest.errors)")
                return nil
            }
        }

        let sortedVersions = versions.sort { (a: LRActionVersion, b: LRActionVersion) -> Bool in
            return (a.primaryVersion.compare(b.primaryVersion) == .OrderedAscending)
        }

        return sortedVersions
    }

    private func _computeAvailableVersionSpecs() -> [LRVersionSpec] {
        var specs: [LRVersionSpec] = []
        var set = Set<LRVersionSpec>()

        for actionVersion in versions {
            let versionSpecs = [
                LRVersionSpec.stableVersionSpecWithMajorFromVersion(actionVersion.primaryVersion),
                LRVersionSpec(matchingMajorMinorFromVersion: actionVersion.primaryVersion),
                LRVersionSpec(matchingVersion: actionVersion.primaryVersion)
            ]
            for spec in versionSpecs {
                if !set.contains(spec) {
                    spec.changeLogSummary = actionVersion.manifest.changeLogSummary
                    set.insert(spec)
                    specs.append(spec)
                }
            }
        }
        
        if let primaryVersionSpace = action.primaryVersionSpace {
            specs.append(LRVersionSpec.stableVersionSpecMatchingAnyVersionInVersionSpace(primaryVersionSpace))
        }
        
        return specs
    }

    private let updateCoalescence = Coalescence()
    private var o = Observation()

}
